src/handlers/miner_hbbft_handler.erl

```diff
@@ -457,77 +478,189 @@ restore(OldState, NewState) ->
+    {opt({gossip, sigs()} | {done, sigs()}), #state{}}.
+state_sigs_check_if_enough(#state{artifact=undefined}=S) ->
+    {none, S};
+state_sigs_check_if_enough(#state{sig_phase = sig, sigs_valid = Sigs, f = F}=S) when length(Sigs) < F + 1 ->
+    {none, S};
+state_sigs_check_if_enough(#state{sig_phase = done, sigs_valid = SigsValid}=S) ->
+    {{some, {done, SigsValid}}, S};
+state_sigs_check_if_enough(
+    #state{
+        sig_phase = Phase,
+        artifact = <<_/binary>>,
+        f = F,
+        sigs_valid = SigsValidAll,
+        signatures_required = Threshold0
+    }=S0
+) ->
+    Threshold1 =
+        case Phase of
+            sig -> F + 1;
+            gossip -> Threshold0
+        end,
+    SigsValidRand =
+        lists:sublist(blockchain_utils:shuffle(SigsValidAll), Threshold1),
+    case
+        length(SigsValidAll) =< (3*F)+1 andalso
+        length(SigsValidRand) >= Threshold1
+    of
+        true ->
+            %% So, this is a little dicey, if we don't need all N signatures,
+            %% we might have competing subsets depending on message order.
+            %% Given that the underlying artifact they're signing is the same
+            %% though, it should be ok as long as we disregard the signatures
+            %% for testing equality but check them for validity
+            SigsValidRandSubset =
+                lists:sublist(lists:sort(SigsValidRand), Threshold0),
             case Phase of
                 gossip ->
-                    {ok, done, lists:sublist(lists:sort(ValidSignatures), Threshold0)};
+                    {{some, {done, SigsValidRandSubset}}, S0};
                 sig ->
-                    case length(ValidSignatures) >= Threshold0 of
+                    case length(SigsValidRand) >= Threshold0 of
                         true ->
-                            {ok, done, lists:sublist(lists:sort(ValidSignatures), Threshold0)};
+                            {{some, {done, SigsValidRandSubset}}, S0};
                         false ->
-                            {ok, gossip, lists:sort(ValidSignatures)}
+                            {{some, {gossip, lists:sort(SigsValidRand)}}, S0}
                     end
             end;
         false ->
-            false
+            {none, S0}
     end.
 
-filter_signatures(State=#state{artifact=Artifact, signatures=Signatures, members=Members}) ->
-    FilteredSignatures = lists:filter(fun({Address, Signature}) ->
-                         lists:member(Address, Members) andalso
-                         libp2p_crypto:verify(Artifact, Signature, libp2p_crypto:bin_to_pubkey(Address))
-                 end, Signatures),
-    State#state{signatures=FilteredSignatures}.
+sig_is_valid(
+    {<<Addr/binary>>, <<Sig/binary>>},
+    #state{artifact=Art, members=Mems, hbbft=HBBFT}
+) ->
+    IsVerified = libp2p_crypto:verify(Art, Sig, libp2p_crypto:bin_to_pubkey(Addr)),
+    IsMember = lists:member(Addr, Mems),
+    IsValid = IsMember andalso IsVerified,
```

Comment:
I'd recommend doing something like:

`IsValid = lists:member(Addr, Mems) andalso libp2p_crypto:verify(Art, Sig, libp2p_crypto:bin_to_pubkey(Addr)),`

This way we don't bother trying to do an expensive signature check if the address is not even one we know about.

Reply:
Makes sense. I did it this way to avoid recomputing `IsVerified` in the negative logging case below.